### ISSUE TITLE
OpenBSD 6.6: cannot allocate memory

### ISSUE THREAD

**Issue by mikdusan** [ziglang/zig#4803]
##### [in case someone else runs into similar issue]

When building binaries (seemingly when they are large) a value in ELF section `OPENBSD_RANDOM` exceeds a kernel limit. According to attached report it is a clang bug and it makes `zig0` fail to execute:
```
cannot allocate memory: zig0
```

The same error is produced running any `llvm-10/bin/` executables if you roll your own for zig.

http://openbsd-archive.7691.n7.nabble.com/execve-1-errno-12-Cannot-allocate-memory-td376879.html

I'm too lazy to put in the work upgrading to OpenBSD `-current` so I'll just wait for the next expected release ~April/2020.

**Comment by mikdusan** [comment]
incidentally, simple patch to get `zig0` building:
```diff
diff --git a/src/os.cpp b/src/os.cpp
index 351b61cd6..102f80408 100644
--- a/src/os.cpp
+++ b/src/os.cpp
@@ -52,7 +52,7 @@ typedef SSIZE_T ssize_t;
 
 #endif
 
-#if defined(ZIG_OS_LINUX) || defined(ZIG_OS_FREEBSD) || defined(ZIG_OS_NETBSD) || defined(ZIG_OS_DRAGONFLY)
+#if defined(ZIG_OS_LINUX) || defined(ZIG_OS_FREEBSD) || defined(ZIG_OS_NETBSD) || defined(ZIG_OS_OPENBSD) || defined(ZIG_OS_DRAGONFLY)
 #include <link.h>
 #endif
 
@@ -1818,7 +1818,7 @@ Error os_get_app_data_dir(Buf *out_path, const char *appname) {
 #endif
 }
 
-#if defined(ZIG_OS_LINUX) || defined(ZIG_OS_FREEBSD) || defined(ZIG_OS_NETBSD) || defined(ZIG_OS_DRAGONFLY)
+#if defined(ZIG_OS_LINUX) || defined(ZIG_OS_FREEBSD) || defined(ZIG_OS_NETBSD) || defined(ZIG_OS_OPENBSD) || defined(ZIG_OS_DRAGONFLY)
 static int self_exe_shared_libs_callback(struct dl_phdr_info *info, size_t size, void *data) {
     ZigList<Buf *> *libs = reinterpret_cast< ZigList<Buf *> *>(data);
     if (info->dlpi_name[0] == '/') {
@@ -1829,7 +1829,7 @@ static int self_exe_shared_libs_callback(struct dl_phdr_info *info, size_t size,
 #endif
 
 Error os_self_exe_shared_libs(ZigList<Buf *> &paths) {
-#if defined(ZIG_OS_LINUX) || defined(ZIG_OS_FREEBSD) || defined(ZIG_OS_NETBSD) || defined(ZIG_OS_DRAGONFLY)
+#if defined(ZIG_OS_LINUX) || defined(ZIG_OS_FREEBSD) || defined(ZIG_OS_NETBSD) || defined(ZIG_OS_OPENBSD) || defined(ZIG_OS_DRAGONFLY)
     paths.resize(0);
     dl_iterate_phdr(self_exe_shared_libs_callback, &paths);
     return ErrorNone;
diff --git a/src/os.hpp b/src/os.hpp
index e73e5e3aa..f65376811 100644
--- a/src/os.hpp
+++ b/src/os.hpp
@@ -27,6 +27,8 @@
 #define ZIG_OS_FREEBSD
 #elif defined(__NetBSD__)
 #define ZIG_OS_NETBSD
+#elif defined(__OpenBSD__)
+#define ZIG_OS_OPENBSD
 #elif defined(__DragonFly__)
 #define ZIG_OS_DRAGONFLY
 #else
```

**Comment by mikdusan** [comment]
so `sysupgrade -s` to snapshot, rebuilding llvm10 and zig fixes that issue... 

but now another problem: OpenBSD is just about the only OS that does not support getting absolute path of current-process' executable. grrr....

**Comment by jreut** [comment]
> so `sysupgrade -s` to snapshot, rebuilding llvm10 and zig fixes that issue...

Out of curiosity, how did you go about building LLVM 10? Did you apply any patches? If I'm looking in the right place, it seems the latest LLVM in ports is version 8.

**Comment by mikdusan** [comment]
> Out of curiosity, how did you go about building LLVM 10? Did you apply any patches? If I'm looking in the right place, it seems the latest LLVM in ports is version 8.

Rolled my own build from llvm's github repository and install to some random location like `/opt/llvm-10.0.0` and point zig's `cmake` setup to that location.

I posted the scripts I used in this tiny repo https://github.com/mikdusan/zig.llvm10 and in its readme there is currently a gdrive link for binary tarballs. Weird that openbsd binaries are so much larger, but I haven't had time to find out why.

**Comment by jreut** [comment]
> I posted the scripts I used in this tiny repo

Thanks so much for the link to your script. For posterity: I'm also on -current and I was able to build after installing the libexecinfo port and prepending `CPATH=/usr/local/include` and `LIBRARY_PATH=/usr/local/lib`.

**Comment by mikdusan** [comment]
solved:

This is either directly related to openbsd default ulimit settings or something else was fixed in OpenBSD 6.8 .

To get zig building on OpenBSD 6.8, bring-your-own LLVM and set ulimit for 10 GB memory and maybe more stack for zig stage1 to be able to build `zig1.o` and run larger tests. The default on openbsd is very low, around 1.5 GB.
